py/day3.py

```diff
@@ -0,0 +1,80 @@
+import string
+from pathlib import Path
+from typing import Generator, Iterable
+
+PRIORITY_LOOKUP: dict[str, int] = {
+    c: i for i, c in enumerate(string.ascii_letters, start=1)
+}
+
+
+# Part 1
+
+
+def bisect_string(s: str) -> tuple[str, str]:
+    s = s.rstrip()
+    return s[: len(s) // 2], s[len(s) // 2 :]
+
+
+def ingest_items(input_file: Path) -> Generator[tuple[str, str], None, None]:
+    with open(input_file) as f_in:
+        for line in f_in:
+            yield bisect_string(line)
+
+
+def find_common_character(collections: tuple[str, str]) -> str:
+    s1, s2 = collections
+    return "".join({c for c in s1 if c in s2})
+
+
+# Part 2
+
+
+def grouper(iterable: Iterable[str], n: int) -> Iterable[tuple[str, ...]]:
+    """ "
+    Collect data into non-overlapping fixed-length chunks or blocks
+
+    Example
+        grouper('ABCDEFGHI', 3) --> ABC DEF GHI
+    """
+    # simplified itertools recipe
+    # https://docs.python.org/3/library/itertools.html#itertools-recipes
+    args = [iter(iterable)] * n
+    return zip(*args, strict=True)
+
+
+def ingest_items_2(input_file: Path) -> Generator[tuple[str, str, str], None, None]:
+    with open(input_file) as f_in:
+        for item in grouper(f_in, 3):
+            yield tuple(x.rstrip() for x in item)  # type: ignore[misc]
```

Comment:
Instead of using this `grouper` function I could have just collected groups of 3 lines and then returned them.
```python
def ingest_items_2(input_file: Path) -> Generator[tuple[str, str, str], None, None]:
    group: list[str] = []
    with open(input_file) as f_in:
        for line in f_in:
            group.append(line.rstrip())
            if len(group) == 3:
                yield tuple(group)  # type: ignore[misc]
                group.clear()
    if group:
        raise ValueError(f"Leftovers remain - {group}")
```